design-document.md

```diff
@@ -14,4 +24,205 @@
+
+Prio combines standard linear secret sharing with a new type of
+probabilistically checkable proof (PCP) system, called a fully linear PCP. The
+input-input validation protocol can be described in terms of three main
+algorithms:
+
+1. pf := Prove(x) denotes generation of a proof pf of the validity of input x.
+   This algorithm is executed by the user.
+1. {vf:i} := Query({x:i}, {pf:i}, r) denotes computation of the verification
+   share {vf:i} for input share {x:i} and proof share {pf:i}. This algorithm is
+   executed by each of the aggregators; input r denotes the joint randomness
+   shared by all of the aggregators.
+1. b := Decide({vf}, r) denotes the execution of the decision procedure on input
+   shares {vf} and joint randomness r. The output b is a boolean indicating
+   whether the input is deemed valid. This algorithm is run by the leader.
+
+The values above have following types:
+
+1. Input x is an element of K^n for some integer n.
+1. Proof pf is an element of K^p(n) for some function p.
+1. The joint randomness r is an element of K^u(n) for some function u.
+1. Each verification share {vf:i} is an element of K^v(n) for some function v.
+
+Before the protocol begins, the aggregators agree on joint randomness r and
+designate one of the aggregators as the leader. The protocol proceeds as
+follows:
+
+1. The user runs pf := Prove(x). It splits x and pf into {x} and {pf}
+   respectively and sends ({x:i}, {pf:i}) to aggregator i.
+1. Each aggregator i runs {vf:i} := Query({x:i}, {pf:i}, r) ands sends {vf:i} to
+   the leader.
+1. The leader runs b := Decide({vf}, r) and sends b to each of the aggregators.
+
+If b=1, then each aggregator i adds its input share {x:i} into its share of the
+aggregate. Once all of the inputs have been validated and aggregated, the
+aggregators send their aggregate shares to the leader, who adds them together to
+obtain the final result.
+
+[[TODO: Sketch out the b=1 path.]]
+
+**Proof generation and verification.**
+[[TODO: Describe how to construct proof systems for languages recognized by
+validity circuits with G-gates, a la [BCC+19, Theorem 4.3].]]
+
+**Security parameters.**
+[[TODO: Define completeness, soundness, and honest-verifier zero-knowledge for
+fully linear PCPs and state bounds for [BBC+19, Theorem 4.3]. This bound will
+guide the selection of the field best suited for the data type and
+application.]]
+
+**Consensus protocol.**
+[[TODO: Describe how the aggregators pick the leader and the joint randomness.]]
+
+### Changes to the input-validation protocol
+
+**Coordinating state.**
+The state of the input-validation protocol is maintained by the leader; except
+for aggregation of the input shares, the other aggregators are completely
+stateless. This is accomplished by making the following changes to the core
+protocol.
+
+1. The user sends all of its shares to the leader. To maintain privacy, the user
+   encrypts each (input, proof) share under the public key of the share's
```

Comment:
How does the public key of the other recipients get advertised?

Reply:
That's TBD. I was thinking the user would simply make an HTTP request to each aggregator, but there might be a better way.

Added a TODO below.